### PR TITLE
feat(query): eliminate the allocation

### LINKED ISSUE #7580: feat: eliminate the allocation.
vectorize_with_builder_4_arg can be used to eliminate the allocation.

_Originally posted by @andylokandy in https://github.com/datafuselabs/databend/pull/7564#discussion_r968585114_

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

vectorize_with_builder_4_arg can be used to eliminate the allocation.

Fixes #7580

### CODE REVIEW

**Comment by TCeason** [comment]
check errr is 

```
2022-09-13T05:23:51.3123046Z 
2022-09-13T05:23:51.3123214Z Crate:         quick-xml
2022-09-13T05:23:51.3123314Z Version:       0.23.0
2022-09-13T05:23:51.3123421Z Warning:       yanked
2022-09-13T05:23:51.3123536Z Dependency tree: 
2022-09-13T05:23:51.3123716Z quick-xml 0.23.0
2022-09-13T05:23:51.3123724Z 
2022-09-13T05:23:51.3123889Z Crate:         quick-xml
2022-09-13T05:23:51.3123984Z Version:       0.24.0
2022-09-13T05:23:51.3124089Z Warning:       yanked
2022-09-13T05:23:51.3124167Z Dependency tree: 
2022-09-13T05:23:51.3124283Z quick-xml 0.24.0
2022-09-13T05:23:51.3124288Z 
2022-09-13T05:23:51.3124399Z error: 1 vulnerability found!
2022-09-13T05:23:51.3124502Z warning: 8 allowed warnings found
2022-09-13T05:23:58.4103655Z ##[error]Process completed with exit code 1.
2022-09-13T05:23:58.4158382Z Post job cleanup.
2022-09-13T05:23:58.4191728Z Post job cleanup.
```

The error will fix in https://github.com/datafuselabs/databend/pull/7577

File: src/query/functions-v2/src/scalars/string.rs [hunk 1]
```diff
@@ -135,11 +136,12 @@ pub fn register(registry: &mut FunctionRegistry) {
         },
     );
 
-    registry.register_4_arg::<StringType, NumberType<i64>, NumberType<i64>, StringType, StringType, _, _>(
+    registry.register_passthrough_nullable_4_arg::<StringType, NumberType<i64>, NumberType<i64>, StringType, StringType, _, _>(
```
Comment (andylokandy) [inline src/query/functions-v2/src/scalars/string.rs:139]:
```suggestion
registry.register_passthrough_nullable_4_arg::<StringType, NumberType<u64>, NumberType<u64>, StringType, StringType, _, _>(
        "insert",
        FunctionProperty::default(),
        |_, _, _, _| None,
        vectorize_with_builder_4_arg::<StringType, NumberType<u64>, NumberType<u64>, StringType, StringType>(
            |srcstr, pos, len, substr, output| {
            let pos = pos as usize;
            let len = len as usize;
            if pos < 1 || pos > srcstr.len() {
                output.put_slice(srcstr);
            } else {
                let pos = pos - 1;
                output.put_slice(&srcstr[0..pos]);
                output.put_slice(substr);
                if pos + len < srcstr.len() {
                    output.put_slice(&srcstr[(pos + len)..]);
                }
            }
            output.commit_row();
            Ok(())
        }),
    );
```

**Comment by andylokandy** [comment]
For example, to eliminate the allocation for `lpad`, we can directly put slice into the builder:

```rust
registry.register_passthrough_nullable_3_arg::<StringType, NumberType<u64>, StringType, StringType, _, _>(
        "lpad",
        FunctionProperty::default(),
        |_, _, _| None,
        vectorize_with_builder_3_arg::<StringType, NumberType<u64>, StringType, StringType>(
            |s, pad_len, pad, output| {
                let pad_len = pad_len as usize;
                if pad_len <= s.len() {
                    output.put_slice(&s[..pad_len])
                } else {
                    let mut remain_pad_len = pad_len - s.len();
                    while remain_pad_len > 0 {
                        if remain_pad_len < pad.len() {
                            output.put_slice(&pad[..remain_pad_len]);
                            remain_pad_len = 0;
                        } else {
                            output.put_slice(pad);
                            remain_pad_len -= pad.len();
                        }
                    }
                    output.put_slice(s);
                }
                output.commit_row();
                Ok(())
            }
        ),
    );
```

File: src/query/functions-v2/src/scalars/string.rs (hunk 1 above)
Comment (TCeason) [reply src/query/functions-v2/src/scalars/string.rs:143]: registry.register_passthrough_nullable_4_arg::<StringType, NumberType<i64>, NumberType<i64>, StringType, StringType, _, _>

I think the i64 is right.

insert support like this: insert('Quadratic', -1, 4, 'What')

Comment (andylokandy) [reply src/query/functions-v2/src/scalars/string.rs:143]: Ok. I was just reading document of snowflake that states: 

> [Valid values are between 1 and one more than the length of the string (inclusive).](https://docs.snowflake.com/en/sql-reference/functions/insert.html#:~:text=Valid%20values%20are%20between%201%20and%20one%20more%20than%20the%20length%20of%20the%20string%20(inclusive))

Comment (andylokandy) [reply src/query/functions-v2/src/scalars/string.rs:143]: I'm wondering why not make negative number counting from the end just like substring 🤔

Comment (TCeason) [reply src/query/functions-v2/src/scalars/string.rs:143]:
```sql
root@mysqldb 15:26:28 [(none)]> select version();
+-----------+
| version() |
+-----------+
| 8.0.29    |
+-----------+
1 row in set (0.00 sec)

root@mysqldb 15:26:33 [(none)]> select  insert('Quadratic', -1, 4, 'What');
+------------------------------------+
| insert('Quadratic', -1, 4, 'What') |
+------------------------------------+
| Quadratic                          |
+------------------------------------+
1 row in set (0.00 sec)

```

Maybe it was for mysql compatibility

**Comment by Xuanwo** [comment]
@mergify update

**Comment by Xuanwo** [comment]
@TCeason Thanks for waiting, https://github.com/datafuselabs/databend/pull/7577 has been merged.

File: src/query/functions-v2/src/scalars/string.rs [hunk 2]
```diff
@@ -110,103 +112,113 @@ pub fn register(registry: &mut FunctionRegistry) {
...
+    registry.register_passthrough_nullable_3_arg::<StringType, StringType, StringType, StringType, _, _>(
         "replace",
         FunctionProperty::default(),
         |_, _, _| None,
-        |str, from, to| {
+        vectorize_with_builder_3_arg::<StringType, StringType, StringType, StringType>(
+            |str, from, to, output| {
             let mut buf: Vec<u8> = vec![];
             if from.is_empty() || from == to {
                 buf.extend_from_slice(str);
```
Comment (andylokandy) [inline src/query/functions-v2/src/scalars/string.rs:201]: This should also be put_slice

Comment (andylokandy) [inline src/query/functions-v2/src/scalars/string.rs:201]:
```suggestion
            if from.is_empty() || from == to {
```

**Review by andylokandy — APPROVED** [review]
rest LGTM

**Comment by TCeason** [comment]
13_0001_q1:                                                             [ FAIL ] - result differs with:
--- /workspace/tests/suites/0_stateless/13_tpch/13_0001_q1.result	2022-09-13 09:38:53.611033684 +0000
+++ /workspace/tests/suites/0_stateless/13_tpch/13_0001_q1.stdout	2022-09-13 09:41:28.374663316 +0000
@@ -1,4 +1 @@
-A	F	3774200.0	5320753880	5054096266.68	5256751331.44	25.5375	36002.1[238](https://github.com/datafuselabs/databend/actions/runs/3043658175/jobs/4903464281#step:4:249)	0.0501	147790
-N	F	95257.0	133737795	127132372.65	132286291.22	25.3006	35521.3269	0.0493	3765
-N	O	7679822.0	10823487077	1.028202505913e10	1.069315804735e10	25.5384	35992.3884	0.05	300716
-R	F	3785523.0	5337950526	5071818532.94	5274405503.04	25.5259	35994.0292	0.0499	148301
+ERROR 1105 (HY000) at line 1: Code: 1043, displayText = Aborted query, because the server is shutting down or the query was killed (while in processor thread 4).

rerun the test.

**Comment by TCeason** [comment]
can success on local

```
12_0002_time_travel_show_tables_hist:                                   [ OK ]
12_0004_time_travel_select_at:                                          [ OK ]
13_0000_prepare:                                                        [ OK ]
13_0001_q1:                                                             [ OK ]
13_0002_q2:                                                             [ OK ]
13_0003_q3:                                                             [ OK ]
13_0004_q4:                                                             [ OK ]
13_0005_q5:                                                             [ SKIPPED ] - skip
13_0006_q6:                                                             [ OK ]
13_0007_q7:                                                             [ OK ]
13_0008_q8:                                                             [ OK ]
13_0009_q9:                                                             [ OK ]
13_0010_q10:                                                            [ OK ]
13_0011_q11:                                                            [ OK ]
13_0012_q12:                                                            [ OK ]
13_0013_q13:                                                            [ OK ]
13_0014_q14:                                                            [ OK ]
13_0015_q15:                                                            [ OK ]
13_0016_q16:                                                            [ OK ]
13_0017_q17:                                                            [ OK ]
13_0018_q18:                                                            [ OK ]
13_0019_q19:                                                            [ OK ]
13_0020_q20:                                                            [ OK ]
13_0021_q21:                                                            [ OK ]
13_0022_q22:                                                            [ OK ]
14_0001_clickhouse_http:                                                [ OK ]
14_0002_select_formats:                                                 [ OK ]
14_0003_http_auth:                                                      [ OK ]
14_0004_http_clickhouse_input_format:                                   [ OK ]
14_0006_input_format_nullable:                                          [ OK ]
14_0007_http_clickhouse_input_format_diagnostic:                        [ OK ]
14_0008_tsv_input_format:                                               [ OK ]
20_0005_mysql_auth:                                                     [ OK ]
20_0008_mysql_client_command:                                           [ OK ]
20_0009_format_diagnostic:                                              [ OK ]

```

test bot is not stable?

```
13_0001_q1:                                                             [ FAIL ] - result differs with:
--- /workspace/tests/suites/0_stateless/13_tpch/13_0001_q1.result	2022-09-13 10:03:04.660878516 +0000
+++ /workspace/tests/suites/0_stateless/13_tpch/13_0001_q1.stdout	2022-09-13 10:05:37.403076707 +0000
@@ -1,4 +1 @@
-A	F	3774200.0	5320753880	5054096266.68	5256751331.44	25.5375	36002.1238	0.0501	147790
-N	F	95257.0	133737795	127132372.65	132286291.22	25.3006	35521.3269	0.0493	3765
-N	O	7679822.0	10823487077	1.028202505913e10	1.069315804735e10	25.5384	35992.3884	0.05	300716
-R	F	3785523.0	5337950526	5071818532.94	5274405503.04	25.5259	35994.0292	0.0499	148301
+ERROR 1105 (HY000) at line 1: Code: 1043, displayText = Aborted query, because the server is shutting down or the query was killed (while in processor thread 4).

13_0019_q19:                                                            [ FAIL ] - result differs with:
--- /workspace/tests/suites/0_stateless/13_tpch/13_0019_q19.result	2022-09-13 10:03:04.664878567 +0000
+++ /workspace/tests/suites/0_stateless/13_tpch/13_0019_q19.stdout	2022-09-13 10:06:06.003514229 +0000
@@ -1 +1 @@
-350370.483
+ERROR 1105 (HY000) at line 2: Code: 1043, displayText = Aborted query, because the server is shutting down or the query was killed (while in processor thread 4).

```

rerun again.

**Pull request merged**